testworld/signature_test.go

```diff
@@ -17,12 +17,55 @@ import (
 	"github.com/centrifuge/go-centrifuge/protobufs/gen/go/document"
 	"github.com/centrifuge/go-centrifuge/testingutils/config"
 	"github.com/centrifuge/go-centrifuge/testingutils/documents"
+	mockdoc "github.com/centrifuge/go-centrifuge/testingutils/documents"
 	"github.com/centrifuge/go-centrifuge/utils"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
 )
 
+func TestHost_SignKeyNotInCollaboration(t *testing.T) {
+	// Hosts
+	alice := doctorFord.getHostTestSuite(t, "Alice")
+	mallory := doctorFord.getHostTestSuite(t, "Mallory")
+
+	actxh := testingconfig.CreateAccountContext(t, alice.host.config)
+
+	// Get PublicKey and PrivateKey
+	publicKey, privateKey := GetSigningKeyPair(t, alice.host.idService, alice.id, actxh)
+	collaborators := [][]byte{mallory.id[:]}
+	dm := createCDWithEmbeddedPO(t, collaborators, alice.id, publicKey, privateKey, alice.host.config.GetContractAddress(config.AnchorRepo))
+
+	//Following simulate attack by Mallory with random keys pair
+	sr, err := dm.CalculateSigningRoot()
+	assert.NoError(t, err)
+	// random keys pairs should cause signature verification failure
+	publicKey2, privateKey2, err := secp256k1.GenerateSigningKeyPair()
+	s, err := crypto.SignMessage(privateKey2, sr, crypto.CurveSecp256K1)
+	assert.NoError(t, err)
+
+	sig := &coredocumentpb.Signature{
+		SignatureId: append(mallory.id[:], publicKey2...),
+		SignerId:    mallory.id[:],
+		PublicKey:   publicKey2,
+		Signature:   s,
+	}
+	malloryDocMockSrv := mallory.host.bootstrappedCtx[documents.BootstrappedDocumentService].(*mockdoc.MockService)
+
+	// so when got request on signature of document, mocking documents.Service of Mallory return a random signature
+	malloryDocMockSrv.On("RequestDocumentSignature", mock.Anything, mock.Anything, mock.Anything).Return(sig, nil).Once()
+
+	malloryDocMockSrv.On("DeriveFromCoreDocument", mock.Anything).Return(dm, nil).Once()
+
+	//TODO
+	signatures, signatureErrors, err := alice.host.p2pClient.GetSignaturesForDocument(actxh, dm)
+	//seems to me, following should get signature verification errors but it is not.  Currenly p2p/client.go just do validateSignatureResp verification (very simple DID verification?), is this the right behavior?
```

Comment:
@zyfrank I believe you are right, at some point we removed the actual signature check (not sure with what purpose).
The block that was removed was:
```
err = identityService.ValidateSignature(resp.Signature, signingRoot)
if err != nil {
	return centerrors.New(code.AuthenticationFailed, fmt.Sprintf("signature invalid with err: %s", err.Error()))
}
```
I believe the test itself is fine, as long as the mocked service is invoked (keep in mind that modifying bootstrapped svcs after the bootstrapping step might not affect to all the other svcs that may had a copy of the original one). 
My suggestion here is to re-add the block above just after this line: https://github.com/centrifuge/go-centrifuge/blob/develop/p2p/client.go#L362
Might need to mock some other unit tests... But at least we will make this test world test fail.

Reply:
@mikiquantum about the mocking service, you are right. Injecting mocking service to test should be done carefully case by case to make sure injecting really happen